File: next-app/models/Car.ts
import mongoose, { Schema, Document } from 'mongoose';

export interface ICar extends Document {
  brand: string;
  model: string;
  year: number;
  price: number;
  bodyType: 'sedan' | 'suv' | 'hatchback' | 'coupe' | 'convertible' | 'wagon' | 'van' | 'truck' | 'luxury';
  fuelType: 'gasoline' | 'diesel' | 'hybrid' | 'electric' | 'plug-in hybrid';
  transmission: 'automatic' | 'manual' | 'semi-automatic' | 'cvt';
  color: string;
  mileage: number;
  engineCapacity: number;
  horsepower?: number;
  features: string[];
  description: string;
  imageUrl: string;
  additionalImages?: string[];
  availability: boolean;
  popularityScore: number;
  createdAt: Date;
  updatedAt: Date;
}

const CarSchema = new Schema<ICar>({
  brand: {
    type: String,
    required: [true, 'Please provide car brand'],
    trim: true
  },
  model: {
    type: String,
    required: [true, 'Please provide car model'],
    trim: true
  },
  year: {
    type: Number,
    required: [true, 'Please provide car year'],
    min: [1900, 'Year must be after 1900'],
    max: [new Date().getFullYear() + 1, 'Year cannot be in the future']
  },
  price: {
    type: Number,
    required: [true, 'Please provide car price'],
    min: [0, 'Price cannot be negative']
  },
  bodyType: {
    type: String,
    required: [true, 'Please provide body type'],
    enum: ['sedan', 'suv', 'hatchback', 'coupe', 'convertible', 'wagon', 'van', 'truck', 'luxury']
  },
  fuelType: {
    type: String,
    required: [true, 'Please provide fuel type'],
    enum: ['gasoline', 'diesel', 'hybrid', 'electric', 'plug-in hybrid']
  },
  transmission: {
    type: String,
    required: [true, 'Please provide transmission type'],
    enum: ['automatic', 'manual', 'semi-automatic', 'cvt']
  },
  color: {
    type: String,
    required: [true, 'Please provide car color']
  },
  mileage: {
    type: Number,
    default: 0
  },
  engineCapacity: {
    type: Number,
    required: [true, 'Please provide engine capacity']
  },
  horsepower: {
    type: Number
  },
  features: {
    type: [String]
  },
  description: {
    type: String,
    required: [true, 'Please provide car description']
  },
  imageUrl: {
    type: String,
    required: [true, 'Please provide car image']
  },
  additionalImages: {
    type: [String]
  },
  availability: {
    type: Boolean,
    default: true
  },
  popularityScore: {
    type: Number,
    default: 0
  }
}, {
  timestamps: true
});

// Add text index for search functionality
CarSchema.index({ brand: 'text', model: 'text', description: 'text' });

// Check if the model exists before defining it to prevent overwriting
export default mongoose.models.Car || mongoose.model<ICar>('Car', CarSchema);

File: fastapi-backend/app/schemas/car.py
from typing import List, Optional
from pydantic import BaseModel, Field
from datetime import datetime
from enum import Enum

class BodyType(str, Enum):
    sedan = "sedan"
    suv = "suv"
    hatchback = "hatchback"
    coupe = "coupe"
    convertible = "convertible"
    wagon = "wagon"
    van = "van"
    truck = "truck"
    luxury = "luxury"

class FuelType(str, Enum):
    gasoline = "gasoline"
    diesel = "diesel"
    hybrid = "hybrid"
    electric = "electric"
    plug_in_hybrid = "plug-in hybrid"

class Transmission(str, Enum):
    automatic = "automatic"
    manual = "manual"
    semi_automatic = "semi-automatic"
    cvt = "cvt"

class CarBase(BaseModel):
    brand: str
    model: str
    year: int = Field(..., ge=1900, le=datetime.now().year + 1)
    price: float = Field(..., ge=0)
    bodyType: BodyType
    fuelType: FuelType
    transmission: Transmission
    color: str
    mileage: int = Field(default=0, ge=0)
    engineCapacity: float
    horsepower: Optional[int] = None
    features: List[str] = []
    description: str
    imageUrl: str
    additionalImages: List[str] = []
    availability: bool = True
    popularityScore: int = 0

class CarCreate(CarBase):
    pass

class CarUpdate(BaseModel):
    brand: Optional[str] = None
    model: Optional[str] = None
    year: Optional[int] = Field(None, ge=1900, le=datetime.now().year + 1)
    price: Optional[float] = Field(None, ge=0)
    bodyType: Optional[BodyType] = None
    fuelType: Optional[FuelType] = None
    transmission: Optional[Transmission] = None
    color: Optional[str] = None
    mileage: Optional[int] = Field(None, ge=0)
    engineCapacity: Optional[float] = None
    horsepower: Optional[int] = None
    features: Optional[List[str]] = None
    description: Optional[str] = None
    imageUrl: Optional[str] = None
    additionalImages: Optional[List[str]] = None
    availability: Optional[bool] = None
    popularityScore: Optional[int] = None

class CarInDB(CarBase):
    id: str
    createdAt: datetime
    updatedAt: datetime

    class Config:
        orm_mode = True

class Car(CarInDB):
    pass

class CarFilter(BaseModel):
    brand: Optional[str] = None
    bodyType: Optional[BodyType] = None
    minPrice: Optional[float] = Field(None, ge=0)
    maxPrice: Optional[float] = Field(None, ge=0)
    fuelType: Optional[FuelType] = None